File: core/src/guardrails/permission.ts
/**
 * core/src/guardrails/permission.ts
 * PermissionRequest フック評価関数
 *
 * permission-request.sh の全ロジックを TypeScript に移植。
 * 安全なコマンド（read-only git、テストコマンド等）を自動承認する。
 *
 * 参照元: scripts/permission-request.sh
 */

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { type HookInput, type HookResult } from "../types.js";

// ============================================================
// PermissionRequest 固有の出力形式
// ============================================================

/** PermissionRequest フックの決定レスポンス */
interface PermissionResponse {
  hookSpecificOutput: {
    hookEventName: "PermissionRequest";
    decision: {
      behavior: "allow" | "deny";
    };
  };
}

function makeAllow(): PermissionResponse {
  return {
    hookSpecificOutput: {
      hookEventName: "PermissionRequest",
      decision: { behavior: "allow" },
    },
  };
}

// ============================================================
// パッケージマネージャー自動承認 許可リスト
// ============================================================

/**
 * .claude/config/allowed-pkg-managers.json が存在し、allowed: true であれば
 * npm/pnpm/yarn の test/build/lint 等を自動承認する。
 */
function isPkgManagerAllowed(cwd: string): boolean {
  const allowlistPath = join(cwd, ".claude", "config", "allowed-pkg-managers.json");
  if (!existsSync(allowlistPath)) return false;

  try {
    const raw = readFileSync(allowlistPath, "utf-8");
    const data = JSON.parse(raw) as unknown;
    if (typeof data === "object" && data !== null && "allowed" in data) {
      return (data as Record<string, unknown>)["allowed"] === true;
    }
  } catch {
    // JSON パースエラーは不許可として扱う
  }

  return false;
}

// ============================================================
// 安全なコマンド判定
// ============================================================

/**
 * コマンド文字列が自動承認可能かを判定する。
 *
 * security hardening:
 * - パイプ、リダイレクト、変数展開、コマンド置換を含む場合は不承認（保守的）
 * - シンプルなコマンドのみ自動承認
 */
function isSafeCommand(command: string, cwd: string): boolean {
  // 複数行コマンドは不承認
  if (command.includes("\n") || command.includes("\r")) return false;

  // シェル特殊文字（パイプ、リダイレクト、変数展開、コマンド置換）を含む場合は不承認
  if (/[;&|<>`$]/.test(command)) return false;

  // read-only git コマンドは常に安全
  if (/^git\s+(status|diff|log|branch|rev-parse|show|ls-files)(\s|$)/i.test(command)) {
    return true;
  }

  // JS/TS テスト・検証コマンドはパッケージマネージャー許可リストを確認
  if (
    /^(npm|pnpm|yarn)\s+(test|run\s+(test|lint|typecheck|build|validate)|lint|typecheck|build)(\s|$)/i.test(
      command
    )
  ) {
    return isPkgManagerAllowed(cwd);
  }

  // Python テスト（package.json リスクなし）
  if (/^(pytest|python\s+-m\s+pytest)(\s|$)/i.test(command)) return true;

  // Go / Rust テスト
  if (/^(go\s+test|cargo\s+test)(\s|$)/i.test(command)) return true;

  return false;
}

// ============================================================
// evaluatePermission: メインエクスポート
// ============================================================

/**
 * PermissionRequest フックの評価関数。
 *
 * Edit/Write は bypassPermissions 相当で自動承認。
 * Bash は安全なコマンドパターンのみ自動承認。
 * その他は何も返さず（デフォルト動作 = ユーザーに確認）。
 */
export function evaluatePermission(input: HookInput): HookResult {
  const toolName = input.tool_name;
  const cwd = input.cwd ?? process.cwd();

  // Edit / Write は自動承認（bypassPermissions モード補完）
  if (toolName === "Edit" || toolName === "Write" || toolName === "MultiEdit") {
    return _permissionResponseToHookResult(makeAllow());
  }

  // Bash 以外はデフォルト動作（スルー）
  if (toolName !== "Bash") {
    return { decision: "approve" };
  }

  // Bash: コマンドを取得して安全性チェック
  const command = input.tool_input["command"];
  if (typeof command !== "string" || command.trim() === "") {
    return { decision: "approve" };
  }

  if (isSafeCommand(command, cwd)) {
    return _permissionResponseToHookResult(makeAllow());
  }

  // 安全でないコマンドはデフォルト動作（ユーザーに確認を委ねる）
  return { decision: "approve" };
}

/**
 * PermissionResponse を HookResult に変換する。
 *
 * PermissionRequest フックは通常の HookResult とは異なる出力形式だが、
 * 内部の型システムでは HookResult として扱い、
 * index.ts の route() で stdout 出力時に formatPermissionOutput() を使って
 * 正しい形式に変換する（Phase 17.1.7 で hooks.json を差し替え後に対応予定）。
 */
function _permissionResponseToHookResult(response: PermissionResponse): HookResult {
  return {
    decision: "approve",
    systemMessage: JSON.stringify(response),
  };
}

/**
 * PermissionRequest フック用の stdout 出力を生成する。
 * index.ts の route() から "permission" フックタイプ時に呼び出す。
 */
export function formatPermissionOutput(result: HookResult): string {
  // systemMessage に PermissionResponse の JSON が入っている場合はそちらを優先
  if (result.systemMessage !== undefined) {
    try {
      const parsed = JSON.parse(result.systemMessage) as unknown;
      if (
        typeof parsed === "object" &&
        parsed !== null &&
        "hookSpecificOutput" in parsed
      ) {
        return JSON.stringify(parsed);
      }
    } catch {
      // パース失敗時は通常の HookResult として出力
    }
  }

  return JSON.stringify(result);
}

File: go/internal/guardrail/permission.go
package guardrail

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"unicode"

	"github.com/tim-hub/powerball-harness/go/pkg/hookproto"
)

// ---------------------------------------------------------------------------
// Package manager allowlist
// ---------------------------------------------------------------------------

func isPkgManagerAllowed(cwd string) bool {
	allowlistPath := filepath.Join(cwd, ".claude", "config", "allowed-pkg-managers.json")
	data, err := os.ReadFile(allowlistPath)
	if err != nil {
		return false
	}
	var obj map[string]interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		return false
	}
	allowed, ok := obj["allowed"]
	if !ok {
		return false
	}
	b, ok := allowed.(bool)
	return ok && b
}

// ---------------------------------------------------------------------------
// Backslash-escape attack detection (CC 2.1.98 bypass mitigation)
// ---------------------------------------------------------------------------

// backslashEscapePattern matches backslash followed by a flag character or space.
// This detects "git\ status", "git\--force", "rm\ -rf" attack vectors.
var backslashEscapePattern = regexp.MustCompile(`\\[\-\s]`)

// hasBackslashEscape returns true if the command contains backslash-escaped
// flag characters or spaces, which are used to bypass pattern matching.
func hasBackslashEscape(cmd string) bool {
	return backslashEscapePattern.MatchString(cmd)
}

// ---------------------------------------------------------------------------
// Env-var prefix bypass detection (CC 2.1.98 bypass mitigation)
// ---------------------------------------------------------------------------

// knownSafeEnvVars is the allowlist of environment variable names that are
// permitted as prefix assignments (e.g. "LANG=C git status").
// Variables prefixed with LC_ (LC_ALL, LC_CTYPE, etc.) are also permitted.
var knownSafeEnvVars = map[string]bool{
	"LANG":        true,
	"LANGUAGE":    true,
	"TZ":          true,
	"NO_COLOR":    true,
	"FORCE_COLOR": true,
}

// isValidEnvVarName checks if s is a valid shell identifier (ASCII letters,
// digits, underscore, must start with letter or underscore).
func isValidEnvVarName(s string) bool {
	if len(s) == 0 {
		return false
	}
	for i, c := range s {
		if i == 0 {
			if !unicode.IsLetter(c) && c != '_' {
				return false
			}
		} else {
			if !unicode.IsLetter(c) && !unicode.IsDigit(c) && c != '_' {
				return false
			}
		}
	}
	return true
}

// stripSafeEnvPrefix strips known-safe "VAR=value" prefix tokens from the
// command and returns the remaining command. Returns ("", false) if any
// unknown environment variable is found in the prefix.
//
// Examples:
//   - "LANG=C git status"       → ("git status", true)
//   - "EVIL=x git status"       → ("", false)
//   - "LANG=C EVIL=x git log"   → ("", false)
//   - "git status"               → ("git status", true)
func stripSafeEnvPrefix(cmd string) (string, bool) {
	tokens := strings.Fields(cmd)
	i := 0
	for i < len(tokens) {
		token := tokens[i]
		eqIdx := strings.IndexByte(token, '=')
		if eqIdx <= 0 {
			// Not a VAR=value token; stop consuming prefix
			break
		}
		varName := token[:eqIdx]
		varValue := token[eqIdx+1:]

		// Reject empty values (VAR= without value)
		if varValue == "" {
			return "", false
		}

		// Reject invalid env var names (must be ASCII identifier)
		if !isValidEnvVarName(varName) {
			return "", false
		}

		// Check against allowlist: known safe vars or LC_ prefix
		if !knownSafeEnvVars[varName] && !strings.HasPrefix(varName, "LC_") {
			return "", false
		}
		i++
	}

	if i == 0 {
		// No prefix tokens consumed; return cmd as-is
		return cmd, true
	}

	// Rejoin remaining tokens
	if i >= len(tokens) {
		// All tokens were env var prefixes, no actual command remains
		return "", false
	}
	return strings.Join(tokens[i:], " "), true
}

// ---------------------------------------------------------------------------
// Safe command patterns
// ---------------------------------------------------------------------------

var (
	safeGitPattern  = regexp.MustCompile(`(?i)^git\s+(status|diff|log|branch|rev-parse|show|ls-files)(\s|$)`)
	safePkgPattern  = regexp.MustCompile(`(?i)^(npm|pnpm|yarn)\s+(test|run\s+(test|lint|typecheck|build|validate)|lint|typecheck|build)(\s|$)`)
	safePytest      = regexp.MustCompile(`(?i)^(pytest|python\s+-m\s+pytest)(\s|$)`)
	safeGoRustTests = regexp.MustCompile(`(?i)^(go\s+test|cargo\s+test)(\s|$)`)
	shellSpecials   = regexp.MustCompile(`[;&|<>` + "`" + `$]`)
)

func isSafeCommand(command, cwd string) bool {
	// Backslash-escaped flags/spaces are an attack vector; reject immediately.
	if hasBackslashEscape(command) {
		return false
	}

	// Strip known-safe env-var prefixes. Unknown prefixes (e.g. EVIL=x) are rejected.
	stripped, ok := stripSafeEnvPrefix(command)
	if !ok {
		return false
	}
	// Evaluate the rest of the command against safe patterns.
	command = stripped

	// Multiline commands are not safe
	for _, c := range command {
		if c == '\n' || c == '\r' {
			return false
		}
	}

	// Shell special chars disqualify
	if shellSpecials.MatchString(command) {
		return false
	}

	// Read-only git is always safe
	if safeGitPattern.MatchString(command) {
		return true
	}

	// JS/TS test/build commands require package manager allowlist
	if safePkgPattern.MatchString(command) {
		return isPkgManagerAllowed(cwd)
	}

	// Python test
	if safePytest.MatchString(command) {
		return true
	}

	// Go / Rust test
	if safeGoRustTests.MatchString(command) {
		return true
	}

	return false
}

// ---------------------------------------------------------------------------
// EvaluatePermission — PermissionRequest hook entry point
// ---------------------------------------------------------------------------

// EvaluatePermission evaluates a PermissionRequest hook.
//
//   - Edit/Write/MultiEdit → auto-allow (bypassPermissions complement)
//   - Bash safe commands → auto-allow
//   - Everything else → pass through (user prompted by CC)
func EvaluatePermission(input hookproto.HookInput) (hookproto.HookResult, *hookproto.PermissionOutput) {
	toolName := input.ToolName
	cwd := input.CWD
	if cwd == "" {
		cwd, _ = os.Getwd()
	}

	// Edit/Write/MultiEdit are auto-allowed
	if matchesWriteEditMultiEdit(toolName) {
		return hookproto.HookResult{Decision: hookproto.DecisionApprove},
			makePermissionAllow()
	}

	// Only Bash gets further evaluation
	if toolName != "Bash" {
		return hookproto.HookResult{Decision: hookproto.DecisionApprove}, nil
	}

	command, ok := input.ToolInput["command"].(string)
	if !ok || command == "" {
		return hookproto.HookResult{Decision: hookproto.DecisionApprove}, nil
	}

	if isSafeCommand(command, cwd) {
		return hookproto.HookResult{Decision: hookproto.DecisionApprove},
			makePermissionAllow()
	}

	// Unsafe command — pass through to user
	return hookproto.HookResult{Decision: hookproto.DecisionApprove}, nil
}

func makePermissionAllow() *hookproto.PermissionOutput {
	return &hookproto.PermissionOutput{
		HookSpecificOutput: hookproto.PermissionHookSpecific{
			HookEventName: "PermissionRequest",
			Decision: hookproto.PermissionDecisionBehavior{
				Behavior: "allow",
			},
		},
	}
}
